### ISSUE TITLE
Are the traces correct?

### ISSUE THREAD

**Issue by andrei-rusu** [karush17/Deep-Eligibility-Traces#5]
Sorry for disturbing you with a question, as I can see this repo is not yet in a fully released state, but this is one of the few codebases around that implements ET in PyTorch, and you also do it in a peculiar way, so I wanted to ask you this!

So, I am wondering whether the methodology you are using for accumulating gradients is actually correct, as the math seems a bit off? The Sutton & Barto book suggests the traces should track the gradient of Q, while the weights are updated according to ``+delta * ET``. It seems to me that your traces track the gradient of the td error instead, which corresponds to ``-delta * grad(Q)``, and then the weight updates happen according to ``-delta * ET`` (although Adam is used so I'd imagine the updates diverge slightly from that).  Now, if SGD would be used, it seems to me that your implementation's updates would effectively be proportional to  ``-delta^2 * grad(Q)``. In this case, wouldn't negative values be completely lost? You may get extra exploration capabilities initially doing it your way, and the problem of ``delta^2`` may actually alleviate as your Q value estimator becomes better, but mathematically-speaking it does seem a bit off...
Was this your intention? Have you explored how your method compares to tracking the gradients of the Q value directly? I am really interested whether this approach makes sense as in my problem doing it your way seems to help.

Thank you for your time!

**Comment by karush17** [comment]
Hi @andrei-rusu! Thank you for the question. Yes, tracking gradients of TD error is intentional. The reason I chose to do it this way is because the gradient of Q estimator `grad(Q)` has small values which may take time to accumulate. Instead, TD error has a better gradient signal which allows updates to be stable. Here is a summary of how the two updates differ-

Sutton & Barto- `w = w + alpha*delta*ET`
This repo- `w = w - alpha*2*delta*grad(Q)*ET`

The reason why the implemented rule comes out this way is because we take derivatives of this objective `delta^2 * ET`. The update is mathematically different from the book but you can think of the extra terms as preconditioners which stabilize gradient descent.

**Comment by andrei-rusu** [comment]
Thank you very much for the quick reply! I understand where you are coming from, but isn't the weight update actually going to have a ``delta^2`` term? You'll accumulate the grad of the td error, which is ``-2*delta*grad(Q)`` and then you multiply the ET with a delta term again. Won't  ``delta^2`` cancel out any negative delta's, thus favoring situations in which actions with negatively-valued delta's get explored?

**Comment by karush17** [comment]
I don't think there is a second multiplication involved with ET. The trace is multiplied only once, with `delta^2`.

**Comment by andrei-rusu** [comment]
Sorry, I missed the fact that the last multiplication is actually with the _squared_ td error. Let me now summarize the 2 steps:
1. The trace accumulates the gradients of the squared td error (here called ``td_error``), which gives rise to a term ``-2*delta*grad(Q)``.
3. The weights get updated as ``w - alpha* Adam(td_error * ET)``

The end result will probably be some modification of the weights proportional to  ``delta*delta^2=delta^3``. Considering the multiplication with the _squared_ td error, then negative values don't get discared indeed.
Thank you for the help!
